### ISSUE TITLE
Redux state removed before components unmount

### ISSUE THREAD

**Issue by colindcampbell** [microsoft/redux-dynamic-modules#85]
I'm seeing an issue when navigating where, when a DynamicModuleLoader component is being removed, the connected container components are still reading from state after the state has already been removed but before the component has dismounted. This leads to reference errors when matchStateToProps is trying to read state.x.y (cannot read property y of undefined) because state.x was already removed. I know I can check if that property exists first, but there seems to be an underlying issue that's causing state to be removed before the component umounts. Using react-redux 7.0.2, react 16.8.6, and redux-dynamic-modules 3.5.0

**Comment by navneet-g** [comment]
Thanks @colindcampbell , can you share a codesandbox with minimal repro?

**Comment by colindcampbell** [comment]
https://codesandbox.io/s/reactredux-uv4xm

**Comment by colindcampbell** [comment]
@navneet-g Have you had any chance to look into this? Thanks

**Comment by ryanf5** [comment]
@colindcampbell 4.0.0 alpha1 has `retained: true` option for your getCountModule, then you register in your createStore: https://codesandbox.io/s/reactredux-z2zkl

**Comment by navneet-g** [comment]
@colindcampbell  sorry I could not look at it as a bit busy in my day job. I will take a look but no promise. Can you upgrade to latest 4.x version and see if this still repros?

**Comment by zulu134** [comment]
@navneet-g  I'm also seeing this. I'm not familiar enough with the code base but it looks like the DynamicModuleLoader might be unmounting before the child component (based on componentWillUnmount on the DML removing the modules which removes the state?), but I don't know if that is even possible. I'm using the saga extension; not sure if that is relevant but I didn't have this problem when using thunks....

**Comment by navneet-g** [comment]
Thanks @zulu134  the state removal from the tree should be the very last thing happens, and by this time any children of DML should be already unmounted. Would you mind debugging and checking who is calling mapStateToProps after DML is unmounted?

**Comment by SantoJambit** [comment]
We're having the issue as well. Not sure if the code by @denisborovnev fixes this entirely, as I tried patching the node_modules file and it didn't seem to work, but maybe I made a mistake.

Our current workaround is like this:

```typescript
/* eslint-disable @typescript-eslint/no-explicit-any */
import { Connect, connect } from 'react-redux';
import { ISagaModule } from 'redux-dynamic-modules-saga';

export function moduleConnect(module: ISagaModule<any>) {
    const keys = module.reducerMap ? Object.keys(module.reducerMap) : [];
    const connect2: Connect = (...args: any[]) => {
        const newArgs = args.map((arg, index) => {
            if (index === 0) {
                return (...args2: any) => {
                    if (keys.every(key => !!args2[0][key])) {
                        return arg(...args2);
                    }
                    return {};
                };
            }
            return arg;
        });
        // @ts-ignore
        return connect(...newArgs);
    };
    return connect2;
}

// Elsewhere:
export default moduleConnect(myModule())(mapStateToProps)(MyComponent);

```

What this does (in a hacky way):
It wraps the mapStateToProps with a function that first checks if all keys in reducerMap are present and if not, it returns an empty object. Since the component itself is never rendered with this value, it should never be an issue (mapStateToProps gets called, but the result of it is never used).

Hoping for an official fix, so we can remove this hack.

Thanks for the great work btw, we love redux-dynamic-modules!

**Comment by abettadapur** [comment]
Published 5.0.2
